## Lazynwb

In [1_polars.ipynb](1_polars.ipynb), we saw predicate pushdown in action to reduce the amount of
data polars reads from a parquet file:

scan parquet file -> `LazyFrame` -> write query -> `.collect()` -> polars optimizes query &
fetches data -> `DataFrame`

The `lazynwb` package allows us to work with NWB files in a similar way.

It works with single files, or multiple files, virtually concatenating tables across files:

In [9]:
import pathlib

import lazynwb
import polars as pl


nwb_paths = list(pathlib.Path('data/dynamicrouting_datacube_v0.0.268/nwb').glob('*.nwb'))
print(f"Found {len(nwb_paths)} NWB files")

Found 252 NWB files


To see the available paths within one NWB file:

In [10]:
lazynwb.get_internal_paths(nwb_paths[0])

{'/acquisition/frametimes_eye_camera/timestamps': <HDF5 dataset "timestamps": shape (267399,), type "<f8">,
 '/acquisition/frametimes_front_camera/timestamps': <HDF5 dataset "timestamps": shape (267204,), type "<f8">,
 '/acquisition/frametimes_side_camera/timestamps': <HDF5 dataset "timestamps": shape (267374,), type "<f8">,
 '/acquisition/lick_sensor_events/data': <HDF5 dataset "data": shape (2734,), type "<f8">,
 '/acquisition/lick_sensor_events/timestamps': <HDF5 dataset "timestamps": shape (2734,), type "<f8">,
 '/intervals/aud_rf_mapping_trials': <HDF5 group "/intervals/aud_rf_mapping_trials" (10 members)>,
 '/intervals/epochs': <HDF5 group "/intervals/epochs" (9 members)>,
 '/intervals/performance': <HDF5 group "/intervals/performance" (21 members)>,
 '/intervals/trials': <HDF5 group "/intervals/trials" (48 members)>,
 '/intervals/vis_rf_mapping_trials': <HDF5 group "/intervals/vis_rf_mapping_trials" (12 members)>,
 '/processing/behavior/dlc_eye_camera': <HDF5 group "/processing/

Access one of the paths as a `LazyFrame` across all NWB files:

- note: `infer_schema_length` optionally limits how many files are scanned to get the schema of the table, since in this case the schema is the same across all files

In [11]:
lf = lazynwb.scan_nwb(nwb_paths, table_path='/intervals/performance', infer_schema_length=1)
lf

We can examine the schema of the table at this point:

In [12]:
lf.collect_schema()

Schema([('aud_dprime', Float64),
        ('aud_nontarget_response_rate', Float64),
        ('aud_target_response_rate', Float64),
        ('block_index', Int64),
        ('catch_response_rate', Float64),
        ('cross_modality_dprime', Float64),
        ('false_alarm_rate', Float64),
        ('hit_rate', Float64),
        ('id', Int64),
        ('is_first_block_aud', Boolean),
        ('n_contingent_rewards', Int64),
        ('n_hits', Int64),
        ('n_responses', Int64),
        ('n_trials', Int64),
        ('rewarded_modality', String),
        ('signed_cross_modality_dprime', Float64),
        ('start_time', Float64),
        ('stop_time', Float64),
        ('vis_dprime', Float64),
        ('vis_nontarget_response_rate', Float64),
        ('vis_target_response_rate', Float64),
        ('_nwb_path', String),
        ('_table_path', String),
        ('_table_index', UInt32)])

note: `_nwb_path` and `_table_row_index` are not columns in the NWB table on disk -- they're added by `lazynwb` to identify the source of each row in a table that spans multiple NWBs

Now we can carry out the same query as we did in [1_polars.ipynb](1_polars.ipynb) for parquet files:

In [13]:
lazynwb.scan_nwb(nwb_paths, table_path='/general/subject', infer_schema_length=1).collect_schema()

Schema([('age', String),
        ('date_of_birth', String),
        ('genotype', String),
        ('sex', String),
        ('species', String),
        ('strain', String),
        ('subject_id', String),
        ('_nwb_path', String),
        ('_table_path', String),
        ('_table_index', UInt32)])

In [14]:
lf = (
    lazynwb.scan_nwb(nwb_paths, table_path='/intervals/performance', infer_schema_length=1)

    # metadata containing session kewords and subject ID is spread out in the NWB file so we have to combine the following two tables:
    .join(
        lazynwb.scan_nwb(nwb_paths, table_path='/general', infer_schema_length=1).select('_nwb_path', 'keywords', 'session_id'),
        on='_nwb_path',
        how='left',
    )
    .join(
        lazynwb.scan_nwb(nwb_paths, table_path='/general/subject', infer_schema_length=1).select('_nwb_path', 'subject_id'),
        on='_nwb_path',
        how='left',
    )
    
    .filter(
        pl.col('keywords').list.contains('production'),
        pl.col('signed_cross_modality_dprime').drop_nans().mean().over('session_id') > 2.0, # <-- for each row, only consider the other rows with the same session_id
    )
    .with_columns(
        (pl.col('stop_time') - pl.col('start_time')).alias('block_duration'),
    )
    .select('block_index', 'block_duration', 'signed_cross_modality_dprime', 'subject_id', '_nwb_path')
    .sort('_nwb_path', 'block_index')
)
lf

In [15]:
lf.collect()

Getting multi-NWB /general/subject table:  19%|███████▍                               | 48/252 [00:01<00:08, 24.10NWB/s]Table '/intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/721536_2024-05-16.nwb

Getting multi-NWB /general/subject table:  55%|████████████████████▊                 | 138/252 [00:08<00:13,  8.66NWB/s]

Getting multi-NWB /general/subject table:  64%|████████████████████████▍             | 162/252 [00:10<00:07, 12.18NWB/s]


Getting multi-NWB /general/subject table:  68%|█████████████████████████▊            | 171/252 [00:10<00:05, 15.03NWB/s]

Getting multi-NWB /general/subject table:  71%|██████████████████████████▉           | 179/252 [00:11<00:04, 15.32NWB/s]

Getting multi-NWB /general/subject table:  72%|███████████████████████████▎          | 181/252 [00:11<00:06, 10.78NWB/s]

Getting multi-NWB /general/subject table:  75%|████████████████████████████▋         | 190/252 [00:12<00:04, 13.08NWB/s]Table '/intervals/performance' not found

shape: (6, 5)
┌─────────────┬────────────────┬───────────────────────────┬────────────┬──────────────────────────┐
│ block_index ┆ block_duration ┆ signed_cross_modality_dpr ┆ subject_id ┆ _nwb_path                │
│ ---         ┆ ---            ┆ ime                       ┆ ---        ┆ ---                      │
│ i64         ┆ f64            ┆ ---                       ┆ str        ┆ str                      │
│             ┆                ┆ f64                       ┆            ┆                          │
╞═════════════╪════════════════╪═══════════════════════════╪════════════╪══════════════════════════╡
│ 0           ┆ 603.68899      ┆ 2.224615                  ┆ 703880     ┆ /root/capsule/data/dynam │
│             ┆                ┆                           ┆            ┆ icrout…                  │
│ 1           ┆ 601.10347      ┆ 1.769163                  ┆ 703880     ┆ /root/capsule/data/dynam │
│             ┆                ┆                           ┆            ┆ icrout…                  │
│ 2           ┆ 604.8567       ┆ 2.639312                  ┆ 703880     ┆ /root/capsule/data/dynam │
│             ┆                ┆                           ┆            ┆ icrout…                  │
│ 3           ┆ 602.07098      ┆ 2.102846                  ┆ 703880     ┆ /root/capsule/data/dynam │
│             ┆                ┆                           ┆            ┆ icrout…                  │
│ 4           ┆ 605.50716      ┆ 2.626949                  ┆ 703880     ┆ /root/capsule/data/dynam │
│             ┆                ┆                           ┆            ┆ icrout…                  │
│ 5           ┆ 597.18357      ┆ 1.232238                  ┆ 703880     ┆ /root/capsule/data/dynam │
│             ┆                ┆                           ┆            ┆ icrout…                  │
└─────────────┴────────────────┴───────────────────────────┴────────────┴──────────────────────────┘

Not as fast as accessing the parquet files (which contain the same data) but much faster than looping over NWB files and opening the tables with the pynwb package.

The largest performance gains will be seen for tables that contain array data in columns, for example the `units` table in ecephys NWBs:


In [16]:
lf = lazynwb.scan_nwb(nwb_paths, table_path='/units', infer_schema_length=1)
{k:v for k,v in lf.collect_schema().items() if k in ('waveform_mean', 'waveform_sd', 'spike_times', 'obs_intervals')}

{'obs_intervals': List(Array(Float64, shape=(2,))),
 'spike_times': List(Float64),
 'waveform_mean': Array(Float64, shape=(210, 384)),
 'waveform_sd': Array(Float64, shape=(210, 384))}

The `waveform_mean` and `waveform_sd` columns contain a `[210 x 384]` array in each row, and each row of the `spike_times` column contains a list with thousands of values.

Opening the units table with pynwb would read all of this data into memory, which is slow and wasteful.

With lazynwb we can filter rows based on metadata columns, then select the array data we want for the remaining rows:

In [17]:
(
    lf  
    .filter(
        pl.col('location').str.starts_with('CA1'),
        pl.col('activity_drift') <= 0.01,
        pl.col('amplitude_cutoff') <= 0.1,
        pl.col('presence_ratio') >= 0.7,
        pl.col('isi_violations_ratio') <= 0.5,
        pl.col('decoder_label') != 'noise',
    )
    .select('unit_id', 'location', 'spike_times', '_nwb_path')
    .collect()
)

Getting multi-NWB /units table:   0%|                                                          | 0/252 [00:00<?, ?NWB/s]

Getting multi-NWB /units table: 100%|██████████████████████████████████████████████████| 64/64 [00:22<00:00,  2.84NWB/s]


shape: (149, 4)
┌─────────────────────────┬──────────┬──────────────────────────────┬──────────────────────────────┐
│ unit_id                 ┆ location ┆ spike_times                  ┆ _nwb_path                    │
│ ---                     ┆ ---      ┆ ---                          ┆ ---                          │
│ str                     ┆ str      ┆ list[f64]                    ┆ str                          │
╞═════════════════════════╪══════════╪══════════════════════════════╪══════════════════════════════╡
│ 759434_2025-02-06_D-33  ┆ CA1      ┆ [20.334285, 20.424817, …     ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 7027.…                       ┆ ut…                          │
│ 759434_2025-02-06_D-73  ┆ CA1      ┆ [20.551917, 20.967581, …     ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 7028.…                       ┆ ut…                          │
│ 644864_2023-02-02_C-159 ┆ CA1      ┆ [512.379566, 512.767995, …   ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 515…                         ┆ ut…                          │
│ 726088_2024-06-18_E-96  ┆ CA1      ┆ [19.944689, 19.975622, …     ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 7104.…                       ┆ ut…                          │
│ 737403_2024-09-27_A-231 ┆ CA1      ┆ [79.594146, 129.606797, …    ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 7741…                        ┆ ut…                          │
│ …                       ┆ …        ┆ …                            ┆ …                            │
│ 628801_2022-09-19_C-271 ┆ CA1      ┆ [34.477293, 34.83749, …      ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 4872.1…                      ┆ ut…                          │
│ 628801_2022-09-19_C-656 ┆ CA1      ┆ [221.385049, 221.400315, …   ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 487…                         ┆ ut…                          │
│ 664851_2023-11-16_B-320 ┆ CA1      ┆ [20.63323, 23.156425, …      ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 6691.3…                      ┆ ut…                          │
│ 706401_2024-04-23_C-243 ┆ CA1      ┆ [20.369653, 24.256623, …     ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 7134.…                       ┆ ut…                          │
│ 706401_2024-04-23_C-256 ┆ CA1      ┆ [19.851291, 19.997423, …     ┆ /root/capsule/data/dynamicro │
│                         ┆          ┆ 7135.…                       ┆ ut…                          │
└─────────────────────────┴──────────┴──────────────────────────────┴──────────────────────────────┘

Behind the scenes, multiple operations take place:
1. the columns referenced in the `.filter()` operation are fetched for 252/252 NWB files
2. the filter operation is applied, which excludes some rows and some NWB files entirely
3. the columns requested in the `.select()` operation are fetched just for the filtered rows, which are found in 64/252 NWB files

---



## SQL -> polars -> NWB

For those who prefer to write SQL queries, `lazynwb` implements the following interface (https://docs.pola.rs/api/python/stable/reference/sql/python_api.html#sqlcontext):

In [ ]:
sql_context = lazynwb.get_sql_context(nwb_paths, infer_schema_length=1)

Scanning NWB files: 100%|███████████████████████| 1/1 [00:00<00:00, 89.01file/s]


<SQLContext [tables:20] at 0x7f276c254310>

In [25]:
sql_context.execute(
    """
    SELECT * FROM performance
    """,
    eager=True,
)

Getting multi-NWB intervals/performance table:   0%|                                           | 0/252 [00:00<?, ?NWB/s]

Getting multi-NWB intervals/performance table:  15%|█████                            | 39/252 [00:00<00:01, 126.31NWB/s]Table 'intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/754852_2025-02-13.nwb
Table 'intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/721536_2024-05-16.nwb
Table 'intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/724612_2024-07-25.nwb
Table 'intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/754852_2025-02-11.nwb
Table 'intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/660023_2023-08-08.nwb
Getting multi-NWB intervals/performance table:  60%|███████████████████             | 150/252 [00:01<00:00, 121.24NWB/s]Table 'intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/754852_2025-02-12.nwb
Table 'intervals/performance' not found in data/dynamicrouting_datacube_v0.0.268/nwb/724612_2024-07-22.nwb
Getting mul

shape: (1_311, 24)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ block_ind ┆ cross_mod ┆ stop_time ┆ aud_nonta ┆ … ┆ is_first_ ┆ _nwb_path ┆ _table_pa ┆ _table_i │
│ ex        ┆ ality_dpr ┆ ---       ┆ rget_resp ┆   ┆ block_aud ┆ ---       ┆ th        ┆ ndex     │
│ ---       ┆ ime       ┆ f64       ┆ onse_rate ┆   ┆ ---       ┆ str       ┆ ---       ┆ ---      │
│ i64       ┆ ---       ┆           ┆ ---       ┆   ┆ bool      ┆           ┆ str       ┆ u32      │
│           ┆ f64       ┆           ┆ f64       ┆   ┆           ┆           ┆           ┆          │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 0         ┆ 2.711324  ┆ 1703.1012 ┆ 0.235294  ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 0        │
│           ┆           ┆ 8         ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ out…      ┆           ┆          │
│ 1         ┆ -0.464653 ┆ 2310.2384 ┆ 0.933333  ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 1        │
│           ┆           ┆ 1         ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ out…      ┆           ┆          │
│ 2         ┆ 2.195563  ┆ 2917.7924 ┆ 0.190476  ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 2        │
│           ┆           ┆ 3         ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ out…      ┆           ┆          │
│ 3         ┆ 1.916772  ┆ 3525.4966 ┆ 1.0       ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 3        │
│           ┆           ┆ 4         ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ out…      ┆           ┆          │
│ 4         ┆ 0.833109  ┆ 4135.9197 ┆ 0.352941  ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 4        │
│           ┆           ┆           ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ out…      ┆           ┆          │
│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │
│ 1         ┆ 0.39923   ┆ 3549.1630 ┆ 0.105263  ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 1        │
│           ┆           ┆ 7         ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ out…      ┆           ┆          │
│ 2         ┆ 0.408895  ┆ 4155.7378 ┆ 0.0       ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 2        │
│           ┆           ┆ 4         ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ out…      ┆           ┆          │
│ 3         ┆ 0.801629  ┆ 4763.3301 ┆ 0.263158  ┆ … ┆ false     ┆ /root/cap ┆ intervals ┆ 3        │
│           ┆           ┆           ┆           ┆   ┆           ┆ sule/data ┆ /performa ┆          │
│           ┆           ┆           ┆           ┆   ┆           ┆ /dynamicr ┆ nce       ┆          │
│           ┆           ┆           ┆     

Note: the metadata tables in `general` are currently missing from the SQL interface but will be added soon: 

In [26]:
sql_context.tables()

['aud_rf_mapping_trials',
 'dlc_eye_camera',
 'epochs',
 'eye_tracking',
 'facemap_front_camera',
 'facemap_side_camera',
 'frametimes_eye_camera',
 'frametimes_front_camera',
 'frametimes_side_camera',
 'lick_sensor_events',
 'licks',
 'lp_front_camera',
 'lp_side_camera',
 'performance',
 'quiescent_interval_violations',
 'rewards',
 'running_speed',
 'trials',
 'units',
 'vis_rf_mapping_trials']